### DISCUSSION TITLE
Agocke's Discriminated Unions Proposal

### DISCUSSION THREAD

**Discussion by YairHalberstadt** [dotnet/csharplang#2962] [General]
@agocke added a proposal for discriminated Unions a few hours ago. I wanted to make a few comments, but you can't comment on a proposal so I thought I'd make this issue to discuss.

The proposal is at https://github.com/dotnet/csharplang/blob/main/proposals/rejected/discriminated-unions.md

**Comment by YairHalberstadt** [comment]
### ValueType DUs

Are you considering ValueType DUs to help reduce allocations.

I think rust's design involving a union and a tag should work in c#, as it is possible to specify that fields in a struct should occupy the same location.

### Top Level Members

If the enum is represented as an abstract class, should it be possible to add top level members to the enum?

I think Niko Matsakis suggestions for DU improvements in Rust which I know @agocke has seen, do a good job of explaining the benefits of allowing this. I'll try to link them soon.

**Reply by ziaulhasanhamim** [reply]
It would be better if it's value type by default like rust and swift. There is no reason for it being allocated on the heap.

**Reply by CyrusNajmabadi** [reply]
@ziaulhasanhamim Value types by default could lead to very large values being passed around with large copying costs along with that.

**Reply by ziaulhasanhamim** [reply]
@CyrusNajmabadi I don't think that's a problem. Because most of the unions won't be that big. Also when you are using a reference type in Union that reference type won't be allocated on the stack with the Union. So if you are adding a big string to a Union the union won't face any issues with stack memory because it has to only store a pointer to the string, not the whole string. If you have many fields for a case you can just create an explicit class.
```c#
enum Employee
{
    Manager(ManagerInfo info),
    Woker(WorkerInfo info)
}
class WorkerInfo
{
// Many fields
}
class ManagerInfo
{
// Many fields
}
```

If there are like 5-6 fields there is no need to do this kinda thing. Swift and rust like languages are doing this without any issue. Note that Value Type Unions should use memory overlap. So that it doesn't waste memory. If you see f# value type discriminated unions each field of each case has separate memory. because memory overlapping is not used there. [more info here](https://bartoszsypytkowski.com/writing-high-performance-f-code/#discriminatedunionsasvaluetypes)

**Reply by CyrusNajmabadi** [reply]
> Because most of the unions won't be that big.

I'm not sure why that would necessarily be the case. 

Similar to records I think the default for a language like c# would be too have classes for her default case and structs of the user opts into it.

**Reply by ziaulhasanhamim** [reply]
But what's the exact reason to make it class over struct?

**Reply by CyrusNajmabadi** [reply]
Zero cost passing of data. Standard class semantics that fits most standard types in .net.  Mutating doesn't operate on a copy, etc.  Consistency with records.  etc. etc.

**Reply by ziaulhasanhamim** [reply]
cost for struct passing is a lot less than overheads of class creation and garbage collection. Also, each operation on class causes a dereference of a pointer which is very cheap but still an extra operation. So copying struct is not as costly as you might think. 
> Standard class semantics that fits most standard types in .net

Where do you think struct doesn't fit?

**Reply by CyrusNajmabadi** [reply]
> cost for struct passing is a lot less than overheads of class creation and garbage collection.

It depends on the domain.  As i mentioned, our default is generally classes (esp for all the other reasons i mentioned).  Given the success taking this route for the record space, i imagine we'd follow the same path here rather than doing something needlessly inconsistent.

**Reply by jnm2** [reply]
> Mutating doesn't operate on a copy

The idea of a DU being mutable makes me shudder.

@CyrusNajmabadi Copy by value wouldn't seem inconsistent to me, given that I think of DUs as enums with parameterized members (and currently manually model them as readonly structs).

**Reply by ziaulhasanhamim** [reply]
Being a value type gives many advantages over reference type. Value types are costless to create and cheap to copy. And if the union size is a problem for being value type you can just create a class and hold it in the value type union. So it gives you a lot of flexibility.

**Reply by HaloFour** [reply]
@jnm2 

> The idea of a DU being mutable makes me shudder.

If you think of a DU as just a closed hierarchy then the mutability of the instance doesn't really matter.  I think in most domains that would suffice. 

 But I think struct DUs are critical for DUs like `Maybe<T>` to keep overhead low.

**Reply by CyrusNajmabadi** [reply]
@ziaulhasanhamim i have no problem with there being a value-type option for DUs.  I think you are not understanding my position.  You do not need to convince me that there are benefits to value types.

**Reply by ziaulhasanhamim** [reply]
> But I think struct DUs are critical for DUs like Maybe<T> to keep overhead low.

@HaloFour  I'm with that. Because at the end of the day it should be like an enum that can hold extra information about the case.

> I think you are not understanding my position

@CyrusNajmabadi Yeah I'm getting your point. You are trying to say to be consistent with records. But unions make more sense to be a value type. And let's be honest with the records too when is the last time we used a large-size record? I think we don't use them often. 
```c#
public readonly record struct Person(
  string FirstName,
  string LastName,
  string Surname,
  string Occupation,
  string FathersName,
  string MothersName
)
```
This isn't large at all because those strings are just pointers. This is true about most things you will store in a record(except primitives which are also very small in size). But I can understand the logic behind the records being reference types by default. But why would unions be a reference type just to be consistent with records? Can you give an example where you really really want to use a reference type union over a value type union?

**Reply by CyrusNajmabadi** [reply]
>  Can you give an example where you really really want to use a reference type union over a value type union?

I think i'd likely use a reference type union almost all the time.  I think of value types as something i used when profiling reveals some sort of hot-spot that needs attention, or there will be just a hojillion of them flying around.  For something like DUs i could see them commonly being used for things like long arg or return type chains.  Having all those copies threading deep in/out of the system just seems superfluous.

I honestly don't get what the big deal is here.  If you want to use a struct DU, i'm virtually certain we'd allow that with little muss/fuss.  Like with records, it would likely be a single token addition.  

Or, if it makes you feel better, we can consider making it equally verbose to declare either.  e.g. `enum class` or `enum struct`.

**Reply by ziaulhasanhamim** [reply]
> I think i'd likely use a reference type union almost all the time

But why? What benefits will you get?

> if it makes you feel better. we can consider making it equally verbose to declare either. e.g. enum class or enum struct

What's the meaning of this? We are talking about making it easier and better to use than feeling better

**Reply by CyrusNajmabadi** [reply]
> But why? What benefits will you get?

I mentioned there benefits already. 

> What's the meaning of this? We are talking about making it easier and better to use than feeling better

I genuinely do not understand why it would be such a big deal to have a single extra token here (if they are different).  Or it will be the same number of tokens if we require both to be explicit. 

That seems totally fine to me, and in line with them language up to now.  

I really don't get why this is an issue.  You'll write the decl once and add whichever keyword is sensible for your domain.  Then your reference it wherever you want and it an works out. 

We're literally talking about one potential token out of thousands of lines of the year of the code for declaring and using the DU.  Why is this such a big deal?

**Reply by jnm2** [reply]
> This isn't large at all because those strings are just pointers.

@ziaulhasanhamim That's six 64-bit pointers, so 48 bytes are being copied implicitly every time the struct value moves from place to place. I believe anything over 16 bytes is considered large.

**Comment by YairHalberstadt** [comment]
I linked the blog posts in an earlier comment [here](https://github.com/dotnet/csharplang/issues/113#issuecomment-513532191)

**Comment by 333fred** [comment]
> If the enum is represented as an abstract class, should it be possible to add top level members to the enum?

Yes, we discussed this explicitly during LDM. The intent is that you can declare a regular partial part of enum Type and put anything you want in there. We also talked about ending the enum section with a semicolon (ie Java-style) and letting you put standard class stuff after that as well. We still need to design if/how base constructors would work in these scenarios, but it's very much on the table.

**Comment by Trayani** [comment]
I'm up for value-type unions ~~and more importantly, for arbitrary type~~.
I mean, we can already do most of what is being proposed with interfaces and/or abstract classes.

What we cannot do is to make generic struct types for patterns like: result-or-exception, T-or-T[], possibly even Value-or-Object-with-Value.
We can tinker with `[StructLayout(LayoutKind.Explicit)]` , but it is limited to non-generic value types-only or classes-only. Could we get more support for this attribute?

**Comment by Richiban** [comment]
Really excited about this; Discriminated Unions are the one feature of any language that I sorely miss with C# and cannot really replicate using other means. Like your designs 👍 

I'm puzzled as to why these will differ from other types though in being `partial` by default. Is there any reason, perhaps one that's not listed in the proposal doc, as to why it would be undesirable or onerous to require the partial keyword?

**Comment by Joe4evr** [comment]
> I'm puzzled as to why these will differ from other types though in being `partial` by default. Is there any reason, perhaps one that's not listed in the proposal doc, as to why it would be undesirable or onerous to require the partial keyword?

The use of `partial` in proposals is just a thing done to keep sample code concise, though it makes a lot more sense when you're proposing to add new API to an existing type. I very much doubt it's actually required in the declaration (especially considering that `partial` is purely a design-time feature, there's no such thing in IL) and the grammar even lists it as `'partial'?` with the question mark indicating it's an optional term.

**Comment by agocke** [comment]
Just added the notes from our discussion of this proposal on Wednesday: https://github.com/dotnet/csharplang/issues/2968

re: struct discriminated unions, I think we should do them and there's a natural parallel here, but it's just out of scope for me right now. I'm trying not to design too many things at the same time

**Comment by Richiban** [comment]
@agocke Any thoughts yet on generic discriminated unions?

Since C# doesn't have a bottom type (aka Nothing or Never) it's not going to be possible to have a generic DU with non-generic cases (or cases with a different arity from the base). 

The easy example of this would be either `Option<T>` or a singly linked list.

**Comment by AartBluestoke** [comment]
> Open questions on the LDM: "Is switching against an argument with an enum class static type guaranteed to be constant-time?"  

I don't believe anywhere else the language spec talks about implementation constraints of O(N) vs O(log(N)). For small N, a Logn scaling implementation (or even hotpath optimisation based on frequency of occurrence from profiling) is often quicker than the overhead of guaranteeing a constant time implementation.

eg: https://github.com/dotnet/csharplang/blob/5dd11df6976beb761c68091cda13475f88ecd9a7/proposals/csharp-8.0/patterns.md
has a small section titled "some possible optimisations" as the only limitations on implementation within the runtime (other than re-ordering)

eg: how compilers implement non-enum switch statements with small (or large) numbers of branches. often "guaranteed constant time" involves some type of hash, jump and collision resolution, which would be more effort than log_2(n) nested if's.

**Comment by agocke** [comment]
@Richiban 

I expect you would define Option as essentially

```C#
enum class Option<T>
{
    Some(T t),
    None
}
```

It's true that you don't have a single None instance for all Option types, but I don't think it's that much of a problem. The only drawback I can see is that when mapping from an `Option<T>` to an `Option<U>` you have to explicitly map between Nones, which doesn't seem that bad.

**Comment by Richiban** [comment]
@agocke I thought as much.

Will there be any special type inference support, perhaps with https://github.com/dotnet/csharplang/issues/2926 (Name lookup for simple name when target type known)?

Basically, what I'd really like to avoid is having to write `Option<string>.None` everywhere, especially if it can be target-typed.

In C# `var x = None` is never going to work where a target type is not available, because you need either a bottom type (as previously mentioned, a la Kotlin, Scala etc) or to allow for generic values (a la Haskell and F#) which means that the generic parameter does not need to be inferred immediately, but can be left for later.

**Comment by agocke** [comment]
Right, `var x = None` won't work, but `Option<string> x = None` should work with #2926. It's not quite as short, but in C# we have a principle that we won't infer a type you can't say, and you can't give a variable an open type (`Option<> x`) in C#. Slightly more verbose but not crazy. In addition, if your return type is `Option<string>` you could just return `None` directly.

**Comment by orthoxerox** [comment]
@Richiban we *might* get a bottom type in some later version of C#, #538 has been championed.

@agocke any chance #538 could be re-triaged to the same milestone as DUs?

**Comment by agocke** [comment]
I don't see how a bottom type would help you in this scenario, since classes are invariant.

**Comment by Richiban** [comment]
As I said before, I'm very much looking forward to getting these DUs! There's a lot of work to be done in the future to make them fully featured, however. For example, this is how you properly model an `Option` in Kotlin.

```
public sealed class Option<out T>
    
public object None : Option<Nothing>()
public data class Some<T>(val value: T) : Option<T>()
```

Unfortunately it makes use of quite a number of features that C# doesn't have:

1. **Covariance in classes**. Necessary for an `Option<Banana>` to be assignable to `Option<Fruit>` etc.
2. **Free objects**. See: `None`. An object declaration is both a type and a value, i.e. a singleton. Can participate in inheritance hierarchies, as in this case.
3. **Sealed classes**. Necessary for the compiler to be able to prove completeness.
4. **Bottom type (Nothing)**. Allows the singleton `None`, in conjunction with (1), to be assignable to `Option<T>` for any `T`.

In addition, it uses a record (marked by the `data` keyword) that gives the `Some` class a public property, structural equality etc.

**Comment by agocke** [comment]
> Covariance in classes. Necessary for an Option<Banana> to be assignable to Option<Fruit> etc.

I don't think this is necessary for all use cases. I doubt this will be in V1 regardless, since variance is a part of the CLR type system and this would require also changing the CLR.

**Comment by gafter** [comment]
> Open questions on the LDM: "Is switching against an argument with an enum class static type guaranteed to be constant-time?"

FYI, our current expectation is yes.  More precisely, we expect amortized constant time performance based on runtime support that we are planning in .Net core 5.  Having said that, we cannot prevent someone else from supplying a "bad" runtime implementation that violates these expectations.

**Comment by orthoxerox** [comment]
Linking to #3760 for future reference

**Comment by Richiban** [comment]
I know there might be a number of DU proposals currently open... does anyone know if any of them is championed?

**Reply by HaloFour** [reply]
There's #113, which has been open for a while but from Discord it sounds like members of the team are interested in picking this up soon.

**Comment by Shadowblitz16** [comment]
- Is this going to be done with the enum keyword or the union keyword?
Because I like the enum variant.

 - will the work with any combination of class, record, struct, record struct, read-only struct and read-only record struct?
Because this looks like it could be done with abstract record structs and record struct inheritance.

- For ad hoc Unions will we be able to define them as alias types or implement interfaces for them?
Because defining ( a or b or c or d ) everytime seems encumbersome.

**Reply by mattwar** [reply]
Here is the current working proposal:
https://github.com/dotnet/csharplang/blob/main/proposals/unions.md

Nominal Type Unions are named.  You won't need to repeat all the types.  There will be some types: Union<T1, T2> etc, that you can use in an ad hoc fashion though.  They will work with most kinds of types.

We are currently debating syntax, including option for enum keyword.
